pkg/descheduler/strategies/node_taint.go

```diff
@@ -0,0 +1,157 @@
+			return 0
+		}
+		totalPods := len(pods)
+		for i := 0; i < totalPods; i++ {
+			if maxPodsToEvict > 0 && nodePodCount[node]+1 > maxPodsToEvict {
+				break
+			}
+			if !checkPodsSatisfyTolerations(pods[i], node) {
+				success, err := evictions.EvictPod(client, pods[i], policyGroupVersion, dryRun)
+				if !success {
+					glog.Errorf("Error when evicting pod: %#v (%#v)\n", pods[i].Name, err)
+				} else {
+					nodePodCount[node]++
+					glog.V(1).Infof("Evicted pod: %#v (%#v)", pods[i].Name, err)
+				}
+			}
+		}
+		podsEvicted += nodePodCount[node]
+	}
+	return podsEvicted
+}
+
+// checkPodsSatisfyTolerations checks if the node's taints (NoSchedule) are still satisfied by pods' tolerations.
+func checkPodsSatisfyTolerations(pod *v1.Pod, node *v1.Node) bool {
+	tolerations := pod.Spec.Tolerations
+	taints := node.Spec.Taints
+	if len(taints) == 0 {
+		return true
+	}
+	noScheduleTaints := getNoScheduleTaints(taints)
+	allTolerated := allTaintsTolerated(noScheduleTaints, tolerations)
+	if !allTolerated {
+		glog.V(2).Infof("Not all taints are tolerated after update for Pod %v on node %v", pod.Name, node.Name)
+		return false
+	}
+	return true
+}
+
+// getNoScheduleTaints return a slice of NoSchedule taints from the a slice of taints that it receives.
+func getNoScheduleTaints(taints []v1.Taint) []v1.Taint {
+	result := []v1.Taint{}
+	for i := range taints {
+		if taints[i].Effect == v1.TaintEffectNoSchedule {
+			result = append(result, taints[i])
+		}
+	}
+	return result
+}
+
+//check validity of Toleration and Taint Effect
+func isValidEffect(tolerationEffect v1.TaintEffect, taintEffect v1.TaintEffect) bool {
+	if len(tolerationEffect) > 0 && tolerationEffect != taintEffect {
+		return false
+	}
+	return true
+}
+
+//check validity of Toleration and Taint Key
+func isValidKey(tolerationKey string, taintKey string) bool {
+	if len(tolerationKey) > 0 && tolerationKey != taintKey {
+		return false
+	}
+	return true
+}
+//toleratesTaint returns true if a toleration tolerates a taint, or false otherwise
+func toleratesTaint(toleration *v1.Toleration, taint *v1.Taint) bool {
+
+	if !isValidEffect(toleration.Effect, taint.Effect){
+		return false
+	}
+	if !isValidKey(toleration.Key, taint.Key){
+		return false
+	}
+	switch toleration.Operator {
+	// empty operator means Equal
+	case "", TolerationOpEqual:
+		return toleration.Value == taint.Value
+	case TolerationOpExists:
+		return true
+	default:
+		return false
+	}
+}
+
+// allTaintsTolerated returns true if all are tolerated, or false otherwise.
+func allTaintsTolerated (taints []v1.Taint, tolerations []v1.Toleration) bool {
+	if len(taints) == 0 {
+		return true
+	}
+	if len(tolerations) == 0 && len(taints) > 0 {
+		return false
+	}
+	for i := range taints {
+		tolerated := false
+		for j := range tolerations {
+			if toleratesTaint(&tolerations[j], &taints[i]) {
+				tolerated = true
+				break
+			}
+		}
+		if !tolerated {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
This doesn't look right. You're `break`ing then returning `true` as soon as you find one taint that can be tolerated, that doesn't test if *all* taints are tolerated. This section can be simplified:
```
for i := range taints {
  for j := range tolerations {
    if !toleratesTaint(&tolerations[j], &taints[i]) {
      return false
    }
  }
}
return true
```
No reason to check `if !tolerated` outside the loop then return `false`, because `if !tolerated` would evaluate to `false` in this case anyway, right? And `toleratesTaint` returns the value you're seeking, so you don't need to store that in a variable anyway.

Reply:
Actually, I was completely mistaken and I apologize for that. You are correct in that we can't simple break on the first failure, because we need to check against all tolerations. This section lgtm